src/data/bucket/circle_bucket.js

```diff
@@ -86,25 +87,32 @@ class CircleBucket<Layer: CircleStyleLayer | HeatmapStyleLayer> implements Bucke
         }
 
         for (const {feature, id, index, sourceLayerIndex} of features) {
-            if (this.layers[0]._featureFilter(new EvaluationParameters(this.zoom), feature)) {
-                const geometry = loadGeometry(feature);
-                const sortKey = circleSortKey ?
-                    circleSortKey.evaluate(feature, {}) :
-                    undefined;
-
-                const bucketFeature: BucketFeature = {
-                    id,
-                    properties: feature.properties,
-                    type: feature.type,
-                    sourceLayerIndex,
-                    index,
-                    geometry,
-                    patterns: {},
-                    sortKey
-                };
-
-                bucketFeatures.push(bucketFeature);
-            }
+            const needGeometry = this.layers[0]._featureFilter.needGeometry;
+            const evaluationFeature = {type: feature.type,
+                id: ('id' in feature ? feature.id : null),
+                properties: feature.properties,
+                geometry: needGeometry ? loadGeometry(feature) : []};
```

Comment:
Nitpick, but this code feels quite repetitive — any way we could handle the branching with geometry loading in a cleaner way? Not necessarily in this PR — we could do it as a follow-up, but this is a good candidate for refactoring. Another thing that doesn't feel good is passing `canonical` parameter across the whole code. One way we could try cleaning this up is introduce a new `EvaluationFeature` class that wraps a vector tile feature, handles lazy loading of geometry, and contains `canonical` as state for later use.

Reply:
I will think about it and do it as a follow-up pr. Thank you for the suggestion!